h/groups/views.py

```diff
@@ -101,26 +104,28 @@ def _read_group(request, group):
     """
     url = request.route_url('group_read', pubid=group.pubid, slug=group.slug)
 
-    result = search.search(request,
-                           private=False,
-                           params={"group": group.pubid, "limit": 1000})
-    annotations = [presenters.AnnotationHTMLPresenter(
-                       storage.annotation_from_dict(a))
-                   for a in result['rows']]
-
-    # Group the annotations by URI.
-    # Create a dict mapping the (normalized) URIs of the annotated documents
-    # to the most recent annotation of each document.
-    annotations_by_uri = collections.OrderedDict()
-    for annotation in annotations:
-        normalized_uri = uri.normalize(annotation.uri)
-        if normalized_uri not in annotations_by_uri:
-            annotations_by_uri[normalized_uri] = annotation
-            if len(annotations_by_uri) >= 25:
-                break
-
-    document_links = [annotation.document_link
-                      for annotation in annotations_by_uri.values()]
+    # Find documents that have at least one annotation that:
+    # a. is shared, and
+    # b. is in this group.
+    documents = (
```

Comment:
Ah, for some reason I didn't think to test the query with groups with large numbers of annotations. Funnily enough I was trying to get this query to work with joins instead of subqueries initially precisely because it

SQL query that the original Python/SQLAlchemy code produced:

```sql
SELECT ... FROM
  SELECT ... FROM document, annotation WHERE EXISTS
    SELECT 1 FROM document_uri, annotation
      WHERE document.id = document_uri.document_id
      AND annotation.target_uri_normalized = document_uri.uri_normalized
      AND annotation.shared = true
      AND annotation.groupid = %(groupid_1)s
  ORDER BY annotation.created
  LIMIT %(param_1)s
AS anon_1 LEFT OUTER JOIN document_uri AS document_uri_1 ON anon_1.document_id = document_uri_1.document_id
LEFT OUTER JOIN document_meta AS document_meta_1 ON anon_1.document_id = document_meta_1.document_id
ORDER BY anon_1.annotation_created, document_uri_1.updated DESC, document_meta_1.updated DESC
```

Query that the above alternative with `join()` instead of `any()` produces:

```sql
SELECT ... FROM
  SELECT ... FROM document JOIN (document_uri AS document_uri_2
                                 JOIN document AS document_1 ON document_uri_2.document_id = document_1.id)
                           ON document_1.id = document_uri_2.document_id
                           JOIN annotation ON annotation.target_uri_normalized = document_uri_2.uri_normalized 
  WHERE annotation.groupid = %(groupid_1)s
  AND annotation.shared = true
  ORDER BY annotation.updated DESC 
  LIMIT %(param_1)s)
  AS anon_1 LEFT OUTER JOIN document_uri AS document_uri_1 ON anon_1.document_id = document_uri_1.document_id
  LEFT OUTER JOIN document_meta AS document_meta_1 ON anon_1.document_id = document_meta_1.document_id
  ORDER BY anon_1.annotation_updated DESC, document_uri_1.updated DESC, document_meta_1.updated DESC
```

So as you say the second subquery is replaced by a (double) join.

I was actually trying to write the query this way in the first place but I couldn't figure out how to use the `annotations` relationship in a join, thanks!